shark/src/main/java/shark/internal/PathFinder.kt

```diff
@@ -462,6 +473,53 @@ internal class PathFinder(
     }
   }
 
+  private fun HeapInstance.readAllNonNullFieldsOfReferenceType(): MutableList<HeapField> {
+    // Assigning to local variable to avoid repeated lookup and cast:
+    // HeapInstance.graph casts HeapInstance.hprofGraph to HeapGraph in its getter
+    val hprofGraph = graph
+    var fieldReader: FieldValuesReader? = null
+    val result = mutableListOf<HeapField>()
+    var skipBytesCount = 0
+    for (heapClass in instanceClass.classHierarchy.toList()) {
+      for (fieldRecord in heapClass.readRecord().fields) {
+        if (fieldRecord.type != PrimitiveType.REFERENCE_HPROF_TYPE) {
+          // Skip all fields that are not references. Track how many bytes to skip
+          skipBytesCount += hprofGraph.getRecordSize(fieldRecord)
+        } else {
+          // Initialize field reader if it's not yet initialized. Replaces `lazy` without synchronization
+          if (fieldReader == null) {
+            fieldReader = hprofGraph.createFieldValuesReader(readRecord())
+          }
+
+          // Skip the accumulated bytes offset
+          fieldReader.skipBytes(skipBytesCount)
+          skipBytesCount = 0
+
+          val fieldValue = fieldReader.readValue(fieldRecord) as ReferenceHolder
+          if (!fieldValue.isNull) {
+            val fieldName = heapClass.fieldName(fieldRecord)
+            result.add(HeapField(heapClass, fieldName, HeapValue(hprofGraph, fieldValue)))
```

Comment:
Looking at the call site, it seems that we only use `fieldName` and `fieldValue.value` (object id). So probably don't need to create all this.

Also, would there be a benefit to this being a sequence instead of creating all fields at once?